File: benchmark/trigger_proxy_aggregation.py
#!/usr/bin/env python3
"""
Trigger proxy/supernode aggregation for VEMB / VSIM using many concurrent
redis-cli processes against the same key.

Why this exists:
- single interactive redis-cli request/response is too serial to trigger proxy batching
- this script fans out many concurrent requests so proxy can accumulate them

No third-party dependencies are required.
"""

import argparse
import os
import subprocess
import sys
import time
from concurrent.futures import ThreadPoolExecutor, as_completed


def run_cmd(cmd):
    start = time.time()
    p = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
    end = time.time()
    return {
        "cmd": cmd,
        "rc": p.returncode,
        "stdout": p.stdout.strip(),
        "stderr": p.stderr.strip(),
        "latency_ms": (end - start) * 1000.0,
    }


def build_vemb_cmd(redis_cli, port, key, element, raw):
    cmd = [redis_cli, "-p", str(port), "VEMB", key, element]
    if raw:
        cmd.append("RAW")
    return cmd


def build_vsim_cmd(redis_cli, port, key, query):
    dim = len(query)
    cmd = [redis_cli, "-p", str(port), "VSIM", key, "VALUES", str(dim)]
    cmd.extend(str(x) for x in query)
    cmd.append("WITHSCORES")
    return cmd


def prefill_vectors(redis_cli, port, key, dim, count):
    print(f"[setup] FLUSHALL")
    print(run_cmd([redis_cli, "-p", str(port), "FLUSHALL"])["stdout"])

    print(f"[setup] inserting {count} vectors into key={key}")
    basis = []
    for i in range(dim):
        vec = [0.0] * dim
        vec[i] = 1.0
        basis.append(vec)

    for i in range(count):
        vec = basis[i % len(basis)]
        cmd = [redis_cli, "-p", str(port), "VADD", key, "VALUES", str(dim)]
        cmd.extend(str(x) for x in vec)
        cmd.append(f"item:{i}")
        res = run_cmd(cmd)
        if res["rc"] != 0:
            print(f"[setup] VADD failed: {' '.join(cmd)}", file=sys.stderr)
            print(res["stderr"], file=sys.stderr)
            sys.exit(1)


def main():
    parser = argparse.ArgumentParser(description="Concurrent redis-cli fanout to trigger proxy aggregation")
    parser.add_argument("--redis-cli", default="./src/redis-cli", help="Path to redis-cli")
    parser.add_argument("--port", type=int, default=6391, help="Redis port")
    parser.add_argument("--key", default="myvectors", help="Target vector key")
    parser.add_argument("--mode", choices=["vemb", "vsim"], required=True, help="Benchmark mode")
    parser.add_argument("--concurrency", type=int, default=32, help="Concurrent redis-cli processes")
    parser.add_argument("--requests", type=int, default=128, help="Total requests")
    parser.add_argument("--raw", action="store_true", help="Use RAW for VEMB")
    parser.add_argument("--dim", type=int, default=4, help="Vector dimension for setup/query")
    parser.add_argument("--prefill-count", type=int, default=64, help="How many vectors to insert before running")
    args = parser.parse_args()

    redis_cli = args.redis_cli
    if not os.path.exists(redis_cli):
        print(f"redis-cli not found: {redis_cli}", file=sys.stderr)
        sys.exit(1)

    prefill_vectors(redis_cli, args.port, args.key, args.dim, args.prefill_count)

    jobs = []
    if args.mode == "vemb":
        for i in range(args.requests):
            element = f"item:{i % args.prefill_count}"
            jobs.append(build_vemb_cmd(redis_cli, args.port, args.key, element, args.raw))
    else:
        query = [1.0] + [0.0] * (args.dim - 1)
        for _ in range(args.requests):
            jobs.append(build_vsim_cmd(redis_cli, args.port, args.key, query))

    print(f"[run] mode={args.mode} requests={args.requests} concurrency={args.concurrency} key={args.key}")

    results = []
    wall_start = time.time()
    with ThreadPoolExecutor(max_workers=args.concurrency) as ex:
        futs = [ex.submit(run_cmd, job) for job in jobs]
        for fut in as_completed(futs):
            results.append(fut.result())
    wall_end = time.time()

    ok = sum(1 for r in results if r["rc"] == 0)
    failed = len(results) - ok
    avg_ms = sum(r["latency_ms"] for r in results) / len(results) if results else 0.0
    qps = len(results) / max(wall_end - wall_start, 1e-9)

    print("")
    print("=== Aggregation Trigger Summary ===")
    print(f"successful: {ok}")
    print(f"failed:     {failed}")
    print(f"avg ms/op:  {avg_ms:.2f}")
    print(f"wall qps:   {qps:.2f}")

    if failed:
        print("")
        print("=== Sample Failures ===")
        shown = 0
        for r in results:
            if r["rc"] != 0 or r["stderr"]:
                print("cmd:", " ".join(r["cmd"]))
                if r["stdout"]:
                    print("stdout:", r["stdout"])
                if r["stderr"]:
                    print("stderr:", r["stderr"])
                print("---")
                shown += 1
                if shown >= 5:
                    break

    print("")
    print("Suggested log checks:")
    print("  tail -n 100 /tmp/redis-vemb-test/redis.log")
    print("  tail -n 100 /tmp/redis-vsim-test/redis.log")
    print("Look for batch sizes > 1 in proxy/supernode logs.")


if __name__ == "__main__":
    main()

File: benchmark/trigger_proxy_aggregation.go
package main

import (
	"bufio"
	"bytes"
	"flag"
	"fmt"
	"net"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

type result struct {
	ok        bool
	latency   time.Duration
	response  string
	errString string
}

func buildRESP(args []string) []byte {
	var b bytes.Buffer
	fmt.Fprintf(&b, "*%d\r\n", len(args))
	for _, a := range args {
		fmt.Fprintf(&b, "$%d\r\n%s\r\n", len(a), a)
	}
	return b.Bytes()
}

func readRESP(br *bufio.Reader) (string, error) {
	line, err := br.ReadString('\n')
	if err != nil {
		return "", err
	}
	if len(line) == 0 {
		return "", fmt.Errorf("empty RESP line")
	}
	prefix := line[0]
	switch prefix {
	case '+', '-', ':':
		return strings.TrimSpace(line[1:]), nil
	case '$':
		n, err := strconv.Atoi(strings.TrimSpace(line[1:]))
		if err != nil {
			return "", err
		}
		if n < 0 {
			return "(nil)", nil
		}
		buf := make([]byte, n+2)
		if _, err := br.Read(buf); err != nil {
			return "", err
		}
		return string(buf[:n]), nil
	case '*':
		n, err := strconv.Atoi(strings.TrimSpace(line[1:]))
		if err != nil {
			return "", err
		}
		parts := make([]string, 0, n)
		for i := 0; i < n; i++ {
			p, err := readRESP(br)
			if err != nil {
				return "", err
			}
			parts = append(parts, p)
		}
		return "[" + strings.Join(parts, ", ") + "]", nil
	default:
		return strings.TrimSpace(line), nil
	}
}

func redisCall(addr string, args []string) (string, error) {
	conn, err := net.DialTimeout("tcp", addr, 2*time.Second)
	if err != nil {
		return "", err
	}
	defer conn.Close()
	if err := conn.SetDeadline(time.Now().Add(5 * time.Second)); err != nil {
		return "", err
	}
	if _, err := conn.Write(buildRESP(args)); err != nil {
		return "", err
	}
	return readRESP(bufio.NewReader(conn))
}

func prefill(addr, key string, dim, count int) error {
	if _, err := redisCall(addr, []string{"FLUSHALL"}); err != nil {
		return err
	}
	basis := make([][]string, dim)
	for i := 0; i < dim; i++ {
		vec := make([]string, dim)
		for j := 0; j < dim; j++ {
			if i == j {
				vec[j] = "1"
			} else {
				vec[j] = "0"
			}
		}
		basis[i] = vec
	}

	for i := 0; i < count; i++ {
		args := []string{"VADD", key, "VALUES", strconv.Itoa(dim)}
		args = append(args, basis[i%dim]...)
		args = append(args, fmt.Sprintf("item:%d", i))
		if _, err := redisCall(addr, args); err != nil {
			return err
		}
	}
	return nil
}

func main() {
	host := flag.String("host", "127.0.0.1", "Redis host")
	port := flag.Int("port", 6391, "Redis port")
	key := flag.String("key", "myvectors", "Vector key")
	mode := flag.String("mode", "vemb", "Mode: vemb|vemb-raw|vsim|vsim-withscores")
	concurrency := flag.Int("concurrency", 32, "Concurrent connections")
	requests := flag.Int("requests", 128, "Total requests")
	dim := flag.Int("dim", 4, "Vector dimension")
	prefillCount := flag.Int("prefill-count", 64, "Prefill vector count")
	flag.Parse()

	addr := fmt.Sprintf("%s:%d", *host, *port)

	fmt.Printf("[setup] prefill key=%s dim=%d count=%d\n", *key, *dim, *prefillCount)
	if err := prefill(addr, *key, *dim, *prefillCount); err != nil {
		fmt.Fprintf(os.Stderr, "prefill failed: %v\n", err)
		os.Exit(1)
	}

	var jobs [][]string
	switch *mode {
	case "vemb":
		for i := 0; i < *requests; i++ {
			jobs = append(jobs, []string{"VEMB", *key, fmt.Sprintf("item:%d", i%*prefillCount)})
		}
	case "vemb-raw":
		for i := 0; i < *requests; i++ {
			jobs = append(jobs, []string{"VEMB", *key, fmt.Sprintf("item:%d", i%*prefillCount), "RAW"})
		}
	case "vsim":
		query := []string{"1"}
		for i := 1; i < *dim; i++ {
			query = append(query, "0")
		}
		for i := 0; i < *requests; i++ {
			args := []string{"VSIM", *key, "VALUES", strconv.Itoa(*dim)}
			args = append(args, query...)
			jobs = append(jobs, args)
		}
	case "vsim-withscores":
		query := []string{"1"}
		for i := 1; i < *dim; i++ {
			query = append(query, "0")
		}
		for i := 0; i < *requests; i++ {
			args := []string{"VSIM", *key, "VALUES", strconv.Itoa(*dim)}
			args = append(args, query...)
			args = append(args, "WITHSCORES")
			jobs = append(jobs, args)
		}
	default:
		fmt.Fprintf(os.Stderr, "unknown mode: %s\n", *mode)
		os.Exit(1)
	}

	fmt.Printf("[run] mode=%s requests=%d concurrency=%d key=%s addr=%s\n",
		*mode, *requests, *concurrency, *key, addr)

	sem := make(chan struct{}, *concurrency)
	results := make(chan result, len(jobs))
	var wg sync.WaitGroup

	startWall := time.Now()
	for _, args := range jobs {
		wg.Add(1)
		go func(a []string) {
			defer wg.Done()
			sem <- struct{}{}
			start := time.Now()
			resp, err := redisCall(addr, a)
			lat := time.Since(start)
			<-sem
			if err != nil {
				results <- result{ok: false, latency: lat, errString: err.Error()}
				return
			}
			results <- result{ok: true, latency: lat, response: resp}
		}(args)
	}

	wg.Wait()
	close(results)
	totalWall := time.Since(startWall)

	var ok, fail int
	var totalLat time.Duration
	samples := 0
	var firstFailures []string

	for r := range results {
		totalLat += r.latency
		samples++
		if r.ok {
			ok++
		} else {
			fail++
			if len(firstFailures) < 5 {
				firstFailures = append(firstFailures, r.errString)
			}
		}
	}

	avgMs := 0.0
	if samples > 0 {
		avgMs = float64(totalLat.Microseconds()) / 1000.0 / float64(samples)
	}
	qps := 0.0
	if totalWall > 0 {
		qps = float64(samples) / totalWall.Seconds()
	}

	fmt.Println("")
	fmt.Println("=== Aggregation Trigger Summary ===")
	fmt.Printf("successful: %d\n", ok)
	fmt.Printf("failed:     %d\n", fail)
	fmt.Printf("avg ms/op:  %.2f\n", avgMs)
	fmt.Printf("wall qps:   %.2f\n", qps)
	if len(firstFailures) > 0 {
		fmt.Println("")
		fmt.Println("=== Sample Failures ===")
		for _, f := range firstFailures {
			fmt.Println(f)
		}
	}

	fmt.Println("")
	fmt.Println("Suggested checks:")
	fmt.Println("  tail -n 100 /tmp/redis-vemb-test/redis.log")
	fmt.Println("  tail -n 100 /tmp/redis-vsim-test/redis.log")
	fmt.Println("Look for worker/proxy batch sizes > 1.")
}
